keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,831 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Running Kubernetes inside Rootless Docker/Podman (kind, minikube)
+When Kubernetes is being executed inside Rootless Docker/Podman, the namespaces and cgroups are already configured by Docker/Podman.
+So, basically there is no additional task, but we still have to modify a few lines of kubelet and kube-proxy to ignore
+minor sysctl & rlimit errors. See ["Required changes to Kubernetes"](#required-changes-to-kubernetes).
+
+
+It should be noted that [kind already works with unmodified Kubernetes](https://kind.sigs.k8s.io/docs/user/rootless/),
+but kind currently uses [very dirty hack to mount fake files under `/proc/sys` to avoid hitting sysctl errors.](https://github.com/kubernetes-sigs/kind/commit/85d51d898a5b642034e908d2869f0ad196c7d052#diff-3c55751d83af635109cece495ee2ff38206764a8b95f4cb8f11fc08a5c0ea8dcR105).
+
+
+### Running Kubernetes directly on the host
+The node components need to be executed inside a user namespace along with other namespaces (mount namespace, network namespace, etc.)
+to gain fake-root privileges, mostly for mount and network operations.
+
+To run Rootless Kubernetes directly on the host, [RootlessKit](https://github.com/rootless-containers/rootlesskit) can be used for creating namespaces.
+In a nutshell, RootlessKit is an extended version of [`unshare`](http://man7.org/linux/man-pages/man1/unshare.1.html) for rootless containers.
+RootlessKit has been already adopted by Docker, BuildKit, Usernetes, k3s, and partially by Podman.
+
+All Kubernetes components including CRI runtime, kubelet, kube-proxy, and CNI daemon need to be executed in RootlessKit's namespaces.
+```console
+$ rootlesskit --net=slirp4netns --copy-up=/etc --copy-up=/run --copy-up=/var --pidns --cgroupns --ipcns --utsns -- containerd &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a kubelet ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a kube-proxy ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -a flanneld ... &
+```
+
+#### Paths
+
+Some paths like `/var/log/pods` are hardcoded in Kubernetes and hard to change.
+
+Although these directories are not writable by unprivileged users, Kubernetes does NOT need to be changed to use unprivileged home directories,
+because RootlessKit can bind-mount writable directories on these paths without the root privileges. (`rootlesskit --copy-up=/var`)
+
+#### Network
+The node components need to be executed in RootlessKit's network namespace, because an unprivileged user cannot do privileged operations in the host network namespace.
+As the components are executed inside a network namespace, `NodePorts` are not directly accessible from other hosts.
+
+An external controller should watch changes on `corev1.Service` resources and call [RootlessKit API](https://github.com/rootless-containers/rootlesskit/blob/v0.11.1/pkg/api/openapi.yaml) to set up port forwarding for the node ports.
```

Comment:
I'm a little confused.  The goal is to run kube-proxy as non-root, but then that doesn't work, so you have something ELSE run as root and forward all the traffic?  That seems kind of pointless.  Am I missing something?  Looking at the impl of port-forwarding it seems to be a user-space implementation.

Reply:
No, the port forwarder is rootless too. This port forwarder is needed because kube-proxy cannot listen on ports outside the namespace.